File: lib/src/main/java/com/walmartlabs/components/scheduler/core/BucketSnapshot.java
package com.walmartlabs.components.scheduler.core;

import com.walmartlabs.components.scheduler.entities.Status;

import java.time.ZonedDateTime;
import java.util.BitSet;

import static com.walmartlabs.components.scheduler.core.BucketManager.L;
import static com.walmartlabs.components.scheduler.entities.Status.PROCESSED;
import static java.lang.String.format;

/**
 * Created by smalik3 on 8/6/16
 */
class BucketSnapshot {

    private static final BitSet EMPTY = new BitSet();

    private final ZonedDateTime id;
    private final long count;
    private final BitSet awaiting;
    private final BitSet processing;

    BucketSnapshot(ZonedDateTime id, long count, BitSet awaiting, BitSet processing) {
        this.id = id;
        this.count = count;
        this.awaiting = awaiting;
        this.processing = processing;
    }

    BucketSnapshot(ZonedDateTime id, long count, int shardSize, String status) {
        this.id = id;
        this.count = count;
        this.processing = new BitSet();
        final int shards = shards(shardSize);
        awaiting = count == 0 || PROCESSED.name().equals(status) ? EMPTY : new BitSet(shards);
        if (awaiting != EMPTY) {
            awaiting.set(0, shards);
        }
        if (count == 0)
            L.info(format("bucket: %s => empty, no events", id));
        else if (awaiting == EMPTY)
            L.info(format("bucket: %s => already done", id));
        else
            L.info(format("bucket: %s => has %d events, resulting in %d shards", id, count, shards));
    }

    BitSet awaiting() {
        return awaiting;
    }

    BitSet processing() {
        return processing;
    }

    long count() {
        return count;
    }

    ZonedDateTime id() {
        return id;
    }

    void done(int shard, Status status) {
        processing.clear(shard);
        switch (status) {
            case PROCESSED:
                L.info(format("shard: %s[%s] finished successfully", id, shard));
                awaiting.clear(shard);
                break;
            case ERROR:
                L.info(format("shard: %s[%s] finished with error", id, shard));
                awaiting.set(shard);
                break;
            default:
                throw new IllegalArgumentException("invalid status value: " + status);
        }
    }

    void processing(int shard) {
        awaiting.clear(shard);
        processing.set(shard);
    }

    private int shards(int shardSize) {
        return (int) (count % shardSize == 0 ? count / shardSize : count / shardSize + 1);
    }

    @Override
    public String toString() {
        return "BucketSnapshot{" +
                "bucket=" + id +
                ", count=" + count +
                ", awaiting=" + awaiting +
                ", processing=" + processing +
                '}';
    }
}

File: lib/src/main/kotlin/com/walmartlabs/bigben/core/BucketSnapshot.kt
/*-
 * #%L
 * BigBen:lib
 * =======================================
 * Copyright (C) 2016 - 2018 Walmart Inc.
 * =======================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package com.walmartlabs.bigben.core

import com.walmartlabs.bigben.entities.EventStatus
import com.walmartlabs.bigben.entities.EventStatus.ERROR
import com.walmartlabs.bigben.entities.EventStatus.PROCESSED
import com.walmartlabs.bigben.utils.logger
import java.time.ZonedDateTime
import java.util.*

internal data class BucketSnapshot(val id: ZonedDateTime, val count: Long, val processing: BitSet, val awaiting: BitSet) {

    companion object {
        private val l = logger<BucketSnapshot>()
        private val EMPTY = BitSet()

        fun with(id: ZonedDateTime, count: Long, shardSize: Int, status: EventStatus): BucketSnapshot {
            val shards = (if (count % shardSize == 0L) count / shardSize else count / shardSize + 1).toInt()
            val awaiting = if (count == 0L || PROCESSED == status) EMPTY else {
                BitSet(shards).apply { set(0, shards) }
            }
            when {
                count == 0L -> if (l.isDebugEnabled) l.debug("bucket: {} => empty, no events", id)
                awaiting === EMPTY -> if (l.isDebugEnabled) l.debug("bucket: {} => already done", id)
                else -> {
                    if (l.isDebugEnabled) l.debug("bucket: {} => has {} events, resulting in {} shards", id, count, shards)
                }
            }
            return BucketSnapshot(id, count, BitSet(), awaiting)
        }
    }

    fun processing(shard: Int) = apply { awaiting.clear(shard); processing.set(shard) }

    fun done(shard: Int, status: EventStatus) {
        processing.clear(shard)
        when (status) {
            PROCESSED -> {
                if (l.isInfoEnabled) l.info("shard: {}[{}] finished successfully", id, shard)
                awaiting.clear(shard)
            }
            ERROR -> {
                if (l.isInfoEnabled) l.info("shard: {}[{}] finished with error", id, shard)
                awaiting.set(shard)
            }
            else -> throw IllegalArgumentException("invalid status value: $status")
        }
    }
}
